src/qibo/backends/pytorch.py

```diff
@@ -199,9 +201,15 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
         ud = self.np.conj(eigenvectors).T
         return self.np.matmul(eigenvectors, self.np.matmul(expd, ud))
 
-    def calculate_matrix_power(self, matrix, power):
-        copied = self.to_numpy(self.np.copy(matrix))
-        copied = super().calculate_matrix_power(copied, power)
+    def calculate_matrix_power(
+        self,
+        matrix,
+        power: Union[float, int],
+        precision_singularity: float = 1e-14,
+    ):
+        copied = self.cast(matrix, copy=True)
+        copied = self.to_numpy(copied) if power >= 0.0 else copied.detach()
+        copied = super().calculate_matrix_power(copied, power, precision_singularity)
```

Comment:
This is due to `torch` not having a `fractional_matrix_power` alternative, right? Or is it due to some of the other `linalg` operations? 
I it's the first case, then I'd say you need to copy and cast to `numpy` only for `power > 0.`

Reply:
TBH idk how `detach` works in `torch`, but detaching the original will not affect the gradients? If the answer is no, then yes I can change it.

Reply:
No indeed, detach basically discards the gradients. But I am not sure why do yo need to detach them for `power > 0.`. All the linalg operations you perform in that case should be `torch` compatible and should not have problems with the gradients if I am not mistaken, or am I missing something?

Reply:
some (maybe all?) `numpy` and `scipy` functions do not work on `torch` tensors if `requires_grad=True`. This error shows up on tests all the time. In this case, it is because of `scipy.linalg.fractional_matrix_power`.

Reply:
Sorry, I meant for `power < 0.` This part in practice
``` python
            # negative powers of singular matrices via SVD
            determinant = self.np.linalg.det(matrix)
            if abs(determinant) < precision_singularity:
                U, S, Vh = self.np.linalg.svd(matrix)
                # cast needed because of different dtypes in `torch`
                S = self.cast(S)
                S_inv = self.np.where(
                    self.np.abs(S) < precision_singularity, 0.0, S**power
                )
                return (
                    self.np.linalg.inv(Vh) @ self.np.diag(S_inv) @ self.np.linalg.inv(U)
                )
```

Reply:
The code only enters this section if `power < 0.0` *and* the matrix is singular. So, for non-singular matrices, the execution still goes through the `scipy` function and thus either still needs the `.detach` or needs to check the determinant at that stage. I chose to just detach every tensor because it seemed it'd be less expensive than calculating the determinant, but I didn't rigorously check it.

Reply:
Ah I see you are right sorry, I missed the det check, which you cannot do outside of the `NumpyBackend` method, I mean you could do but it wouldn't make much sense.